### PR TITLE
Add bulk role and policy delete APIs

### LINKED ISSUE #3497: Add bulk delete APIs for ZMS roles and policies
### Background

Large Athenz domains can contain thousands of roles and policies. When ZMS uses `ZMSFileChangeLogStore`, each `deleteRole` or `deletePolicy` operation can cause ZTS to synchronize the affected large domain from ZMS.

In environments with many large domains, deleting roles or policies one at a time can create a high volume of ZTS-to-ZMS synchronization requests.
This can increase ZMS load, cause request backlog, leave ZTS cache data stale while responses are delayed, and eventually make both ZMS and ZTS behavior unstable.
This is commonly referred to as the domain cache problem.

### Problem

ZMS currently supports deleting one role or one policy per request.
Operational tasks that delete hundreds of roles or policies must issue hundreds of individual requests, and each successful request can trigger a separate domain synchronization for large domains.

For example, one domain may have over 10,000 roles. While deleting all of them is unlikely, practical cleanup operations can still involve 200 to 300 roles or policies. The current one-at-a-time API shape makes those cleanup operations much more expensive than necessary.

### Proposal

Add REST DELETE endpoints that accept multiple names in the request path:

```text
DELETE /domain/{domainName}/roles/{roleNames}
DELETE /domain/{domainName}/policies/{policyNames}
```

`roleNames` and `policyNames` are comma-separated lists of entity names.

The server should process each bulk request as a single logical operation:

- Validate all names before applying deletes.
- Preserve existing delete semantics for authorization, audit requirements, resource ownership, and admin role/policy protection.
- Delete all requested roles or policies in one transaction.
- Keep per-object audit log entries.
- Publish one domain-level change message per bulk request so that `ZMSFileChangeLogStore` causes one large domain synchronization per batch instead of one per deleted object.

Client libraries and CLI tooling should also support chunking large delete lists so callers can pass a practical list of names without manually splitting requests. The initial client-side chunking limit should be:

- 250 names per request.
- 6000 characters per encoded path parameter.

This keeps request paths within common proxy/server limits while reducing large-domain synchronization frequency by roughly 1/200 to 1/300 for common cleanup operations.

### Acceptance Criteria

- ZMS exposes bulk role delete and bulk policy delete endpoints using request path parameters, not query parameters or request bodies.
- Java and Go clients expose helpers for bulk role and policy deletes.
- `zms-cli` supports deleting multiple roles or policies and chunks requests automatically.
- Bulk delete requests preserve existing validation and error behavior where practical.
- A failed validation or missing target prevents partial deletion for that request.
- Successful bulk deletes generate one domain-level change message per request.
- Tests cover server behavior, client chunking, CLI behavior, validation failures, and change message behavior.

### PR DESCRIPTION
**Pull request by ctyano**
# Description
Resolves #3497

# Contribution Checklist:
- [x] **The pull request does not introduce any breaking changes**
- [x] **I have read the [contribution guidelines](https://github.com/AthenZ/athenz/blob/master/CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request.**

### CODE REVIEW

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java [hunk 1]
```diff
@@ -5209,20 +5209,97 @@ public void deleteRole(ResourceContext ctx, String domainName, String roleName,
...
+            } catch (ServerResourceException ex) {
+                throw ZMSUtils.error(ex);
             }
+            roles.add(role);
         }
 
-        dbService.executeDeleteRole(ctx, domainName, roleName, auditRef, caller);
+        for (Role role : roles) {
+            deleteTenantAssumeRoleAssertions(ctx, domainName, role, auditRef, caller);
+        }
```
[the remark this answers is not in the document]
Comment (ctyano) [reply servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java:5280]: Fixed. The bulk role delete path now collects tenant assume_role cleanup requests during validation but executes the cleanup inside `DBService.executeDeleteRoles` using the same transaction as the provider role deletes.
Cleanup audit/change messages are emitted after commit, and cleanup failures roll back before role deletion proceeds. Also added tests for cleanup rollback failure and bulk trusted-role cleanup success.

File: clients/go/zms/client.go [hunk 2]
```diff
@@ -1237,6 +1237,37 @@ func (client ZMSClient) DeleteRole(domainName DomainName, roleName EntityName, a
 	}
 }
 
+func (client ZMSClient) DeleteRoles(domainName DomainName, roleNames EntityNameList, auditRef string, resourceOwner string) error {
+	headers := map[string]string{
+		"Athenz-Resource-Owner": resourceOwner,
+		"Y-Audit-Ref":           auditRef,
+	}
+	url := client.URL + "/domain/" + fmt.Sprint(domainName) + "/roles/" + fmt.Sprint(roleNames)
+	resp, err := client.httpDelete(url, headers)
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/go/zms/client.go:1246]: Fixed. Added generator-safe Go client helpers in a separate file: `DeleteRoleList` and `DeletePolicyList`.
They issue bounded requests using the 250-name and 6000-character path parameter limits, with tests. The `zmscli` bulk delete paths now use these public helpers.

File: clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java [hunk 3]
```diff
@@ -114,12 +116,50 @@ public class ZMSClient implements Closeable {
...
+            throw new IllegalArgumentException("no entity names specified");
+        }
+
+        String[] names = entityNames.split(",", -1);
+        StringBuilder chunk = new StringBuilder();
+        int chunkCount = 0;
+
+        for (String name : names) {
+            String entityName = name.trim();
+            int nextLength = chunk.length() + (chunkCount == 0 ? 0 : 1) + entityName.length();
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java:145]: Fixed. The Java ZMS client now rejects any individual entity name that exceeds the 6000-character path parameter limit before chunking.
Added role and policy tests to verify oversized names fail client-side without sending a request.

File: libs/go/zmscli/cli.go [hunk 4]
```diff
@@ -30,8 +30,33 @@ const (
...
+	for start := 0; start < len(names); {
+		end := start
+		pathParamLength := 0
+		for end < len(names) && end-start < bulkDeleteChunkSize {
+			nextLength := len(names[end])
+			if end > start {
+				nextLength++
+			}
+			if end > start && pathParamLength+nextLength > bulkDeleteMaxPathParamLength {
+				break
```
[the remark this answers is not in the document]
Comment (ctyano) [reply libs/go/zmscli/cli.go]: Fixed. The CLI-local chunking was replaced with the shared Go client bulk delete helpers, which reject individually oversized names before issuing requests.
Also added CLI tests for oversized role and policy names.

File: clients/go/zms/client_bulk_delete.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		return nil, fmt.Errorf("no %s names specified", objectType)
+	}
+
+	chunks := make([]EntityNameList, 0, (len(names)+BulkDeleteChunkSize-1)/BulkDeleteChunkSize)
+	chunk := make([]string, 0, BulkDeleteChunkSize)
+	pathParamLength := 0
+
+	for _, name := range names {
+		entityName := string(name)
+		entityNameLength := len(entityName)
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/go/zms/client_bulk_delete.go:35]: Fixed. The Go client chunker now deduplicates names case-insensitively before building request chunks, matching the server-side normalization behavior and avoiding a later chunk failing with NOT_FOUND for a duplicate already deleted by an earlier chunk.
Also added coverage for duplicates across a chunk boundary.

File: clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java [hunk 6]
```diff
@@ -114,12 +116,53 @@ public class ZMSClient implements Closeable {
...
+            throw new IllegalArgumentException("no entity names specified");
+        }
+
+        String[] names = entityNames.split(",", -1);
+        StringBuilder chunk = new StringBuilder();
+        int chunkCount = 0;
+
+        for (String name : names) {
+            String entityName = name.trim();
+            if (entityName.length() > BULK_DELETE_MAX_PATH_PARAM_LENGTH) {
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java]: Fixed. The Java client chunker now tracks normalized names and skips case-insensitive duplicates before issuing any delete requests, matching the server-side deduplication behavior.
Also added role and policy tests for duplicates across a chunk boundary.

File: libs/go/zmscli/role.go [hunk 7]
```diff
@@ -171,6 +171,31 @@ func (cli Zms) DeleteRole(dn string, rn string) (*string, error) {
 	return cli.dumpByFormat(message, cli.buildYAMLOutput)
 }
 
+func (cli Zms) DeleteRoles(dn string, roleNames []string) (*string, error) {
+	if len(roleNames) == 0 {
+		return nil, fmt.Errorf("no role names specified")
+	}
+	for _, rn := range roleNames {
+		if rn == "admin" {
```
[the remark this answers is not in the document]
Comment (ctyano) [reply libs/go/zmscli/role.go]: Fixed. The zmscli role delete pre-validation now uses a case-insensitive `admin` role check, so `Admin` and other case variants are rejected before any chunked delete request is sent.
Updated the CLI test accordingly.

File: clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java [hunk 8]
```diff
@@ -114,12 +119,57 @@ public class ZMSClient implements Closeable {
...
+        String[] names = entityNames.split(",", -1);
+        StringBuilder chunk = new StringBuilder();
+        int chunkCount = 0;
+        Set<String> normalizedNames = new HashSet<>();
+
+        for (String name : names) {
+            String entityName = name.trim();
+            if (entityName.length() > BULK_DELETE_MAX_PATH_PARAM_LENGTH) {
+                throw new IllegalArgumentException("entity name exceeds maximum path parameter length");
+            }
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java]: Fixed. The Java client now validates the full input before sending any delete request, and the entity-name syntax check uses the RDL schema validator (`ZMSSchema.instance()`) in client code. This keeps the naming rule sourced from the schema while still preventing invalid later entries from causing partial chunk deletion.
Also added role and policy tests for an oversized name after a full chunk, plus empty/invalid name coverage.

File: libs/go/zmscli/policy.go [hunk 9]
```diff
@@ -388,6 +388,26 @@ func (cli Zms) DeletePolicy(dn string, pn string) (*string, error) {
 	return cli.dumpByFormat(message, cli.buildYAMLOutput)
 }
 
+func (cli Zms) DeletePolicies(dn string, policyNames []string) (*string, error) {
+	if len(policyNames) == 0 {
+		return nil, fmt.Errorf("no policy names specified")
+	}
```
[the remark this answers is not in the document]
Comment (ctyano) [reply libs/go/zmscli/policy.go:394]: Fixed. The bulk policy delete command now scans the complete policy argument list case-insensitively and rejects the protected `admin` policy before calling `DeletePolicyList`, matching the role bulk delete behavior and avoiding partial deletion when `admin` appears in a later chunk.
Also added CLI test coverage for `Admin`.

File: clients/go/zms/client_bulk_delete.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+			return nil, fmt.Errorf("%s name exceeds maximum path parameter length", objectType)
+		}
+		if val := rdl.Validate(ZMSSchema(), "EntityName", entityName); !val.Valid {
+			return nil, fmt.Errorf("invalid %s name specified", objectType)
+		}
+		normalizedName := strings.ToLower(entityName)
+		if _, ok := normalizedNames[normalizedName]; ok {
+			continue
+		}
+		normalizedNames[normalizedName] = struct{}{}
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/go/zms/client_bulk_delete.go]: Fixed. The Go bulk delete helper now rejects the protected `admin` role/policy case-insensitively during full-list prevalidation, before any chunked request is sent. Also added tests where `Admin` appears after a full chunk to verify no request is dispatched.

File: clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java [hunk 11]
```diff
@@ -114,12 +122,86 @@ public class ZMSClient implements Closeable {
...
+            if (!BULK_DELETE_ENTITY_NAME_VALIDATOR.validate(entityName, "EntityName").valid) {
+                throw new IllegalArgumentException("invalid entity name specified");
+            }
+            if (!normalizedNames.add(entityName.toLowerCase(Locale.ROOT))) {
+                continue;
+            }
+            names.add(entityName);
```
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/java/zms/src/main/java/com/yahoo/athenz/zms/ZMSClient.java]: Fixed. The Java bulk delete helper now rejects `admin` case-insensitively during the full-list prevalidation phase before chunking or dispatch. Also added role and policy tests with `Admin` in a later chunk to verify the generated client is never called.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java [hunk 12]
```diff
@@ -2840,6 +3016,79 @@ void executeDeletePolicy(ResourceContext ctx, String domainName, String policyNa
...
+
+                // add domain change event
+                addDomainChangeMessage(ctx, domainName, domainName, DomainChangeMessage.ObjectType.DOMAIN);
+
+                return;
+
+            } catch (ServerResourceException ex) {
+                if (!shouldRetryOperation(ex, retryCount)) {
+                    throw ZMSUtils.error(ex);
+                }
```
[the remark this answers is not in the document]
Comment (ctyano) [reply servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java:3093]: Fixed. `executeDeletePolicies` now wraps the transactional body in an inner `ServerResourceException` catch and calls `rollbackChanges(con)` before the connection can close, matching the `executeDeleteRoles` pattern. Also added a test where the second policy deletion throws after the first deletion succeeds, verifying rollback with no commit or timestamp update.

File: clients/go/zms/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ctyano) [reply clients/go/zms/client.go:1246]: No further change is needed in the generated one-request methods. The bounded public helpers are `DeleteRoleList` and `DeletePolicyList`, implemented outside the generated file so they are generator-safe. They now perform full-list prevalidation, case-insensitive deduplication, protected `admin` rejection, and 250-name / 6000-character chunking before dispatching requests.

**Comment by havetisyan** [comment]
@ctyano I see a fundamental problem with your logic how to chunk and deletes roles and policies from the client side. Let's assume I want to delete 1000 roles. Based on your client logic it will create chunks (let's assume each chunk is 100 roles) so it fill carry out 10 bulk role delete operations. After the third bulk operation, the request is rejected. So now the operation is rejected yet the client is not told that the first 300 roles were successfully deleted and at this from the client side any type of rollback is not possible. So you have two options:

a) need to keep track of all chunks, which ones were successful which one failed and then you can't really throw an exception back to the client since you technically can return partial success response, etc which doesn't really fit with the rest api that we have in place

b) don't provide any type of chunking and let the admin handle the exceptions.  In Athenz we already provide the capability to set the max request size for jetty (requestHeaderSize). So if you want, you can increase that limit for your server. If your client generates a very long size, I believe jetty should reject it with 431 error code (this could get mapped to bad request from client side) but the message is there so your admin can decide to limit the size.

**Comment by ctyano** [comment]
@havetisyan You're right. I made the client-side behavior more feature rich than this API needs to be. Option (b) is the appropriate direction here.

Client-side chunking would turn a single logical bulk delete into multiple independent server requests, which can result in partial completion if a later request fails. I have removed the client-side chunking so each client/CLI operation maps to a single server-side bulk delete request and preserves the API's all-or-fail semantics. Oversized requests are now left to the existing server/Jetty request size configuration, as suggested.
